crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,878 @@
+            active_authority.net.clone(),
+            timing.extra_time_after_quorum,
+            timing.timeout_until_quorum,
+        )
+        .await;
+
+        if let Err(err) = state_of_world {
+            warn!("Cannot get a quorum of checkpoint information: {:?}", err);
+            // Sleep for 10 sec to allow the network to set itself up or the partition
+            // to go away.
+            tokio::time::sleep(timing.delay_on_quorum_failure).await;
+            continue;
+        }
+
+        let (checkpoint, proposals) = state_of_world.expect("Just checked that we are not Err");
+
+        // (2) Sync to the latest checkpoint, this might take some time.
+        // Its ok nothing else goes on in terms of the active checkpoint logic
+        // while we do sync. We are in any case not in a position to make valuable
+        // proposals.
+        if let Some(checkpoint) = checkpoint {
+            // Check if there are more historic checkpoints to catch up with
+            let next_checkpoint = state_checkpoints.lock().next_checkpoint();
+            if next_checkpoint < checkpoint.checkpoint.sequence_number {
+                // TODO log error
+                if let Err(err) = sync_to_checkpoint(
+                    active_authority.state.name,
+                    active_authority.net.clone(),
+                    state_checkpoints.clone(),
+                    checkpoint.clone(),
+                )
+                .await
+                {
+                    warn!("Failure to sync to checkpoint: {}", err);
+                    // if there was an error we pause to wait for network to come up
+                    tokio::time::sleep(timing.delay_on_quorum_failure).await;
+                }
+
+                continue;
+            }
+
+            // Check if the checkpoint is the one we are expecting next!
+            // if next_checkpoint == checkpoint.checkpoint.sequence_number {
+            // Try to upgrade the signed checkpoint to a certified one if possible
+            if state_checkpoints
+                .lock()
+                .handle_checkpoint_certificate(&checkpoint, &None)
+                .is_err()
+            {
+                // One of the errors may be due to the fact that we do not have
+                // the full contents of the checkpoint. So we try to download it.
+                // TODO: clean up the errors to get here only when the error is
+                //       "No checkpoint set at this sequence."
+                if let Ok(contents) = get_checkpoint_contents(
+                    active_authority.state.name,
+                    active_authority.net.clone(),
+                    &checkpoint,
+                )
+                .await
+                {
+                    // Retry with contents
+                    let _ = state_checkpoints
+                        .lock()
+                        .handle_checkpoint_certificate(&checkpoint, &Some(contents));
```

Comment:
This is a very good question, and I am not sure we need this code in `handle_checkpoint_certificate`:
```
// Now that we have the new checkpoint we try to move forward the checkpoint creation
                    // process. We try to use fragments in the sequence to create past checkpoints.
                    loop {
                        let construct = self.attempt_to_construct_checkpoint();
                        // Exit if checkpoint construction leads to an error or returns false
                        // (ie no new checkpoint is created.)
                        if construct.is_err() || !construct.unwrap() {
                            break;
                        }
                    }
```
This code is attempting to construct checkpoints from fragments in the sequence for subsequent checkpoints. The ideas is that if a node falls behind maybe there are are many checkpoints that can be created through fragments, not just one. However, even if the fragments are assembled they are unlikely to form a checkpoint that this node can decode, since it has made no proposal and therefore there will be no fragment connecting its (non existing) proposal to the others.

So not sure if this code makes much sense.